### PR TITLE
PHPStan (Level 4) on controllers/

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | PHPStan (Level 4) on controllers/
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Relative to #16471
| How to test?      | Github Actions are :green_circle: & @boubkerbribri are :green_circle: 

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27102)

### CODE REVIEW

File: classes/Store.php [hunk 1]
```diff
@@ -59,7 +59,7 @@ class StoreCore extends ObjectModel
     /** @var float Longitude */
     public $longitude;
 
-    /** @var string Store hours (PHP serialized) */
+    /** @var string|array Store hours (PHP serialized) */
```
Comment (PierreRambaud) [inline classes/Store.php:62]: How it could be an array?

File: controllers/admin/AdminAccessController.php [hunk 2]
```diff
@@ -137,6 +137,7 @@ public function initPageHeaderToolbar()
 
     public function ajaxProcessUpdateAccess()
     {
+        /* @phpstan-ignore-next-line */
```
Comment (PierreRambaud) [inline controllers/admin/AdminAccessController.php:140]: Why? the `_PS_MODE_DEMO_` is defined in defines.inc.php :thinking:

File: controllers/admin/AdminCustomerThreadsController.php [hunk 3]
```diff
@@ -1171,9 +1171,7 @@ public function syncImap()
                         }
                     }
                 }
-                if ($fetch_succeed) {
-                    Db::getInstance()->execute('INSERT INTO `' . _DB_PREFIX_ . 'customer_message_sync_imap` (`md5_header`) VALUES (\'' . pSQL($md5) . '\')');
-                }
+                Db::getInstance()->execute('INSERT INTO `' . _DB_PREFIX_ . 'customer_message_sync_imap` (`md5_header`) VALUES (\'' . pSQL($md5) . '\')');
```
Comment (PierreRambaud) [inline controllers/admin/AdminCustomerThreadsController.php:1171]: `fetch_succeed` can be false because of line 1167 :thinking:

File: controllers/admin/AdminLegacyLayoutController.php [hunk 4]
```diff
@@ -197,7 +197,7 @@ public function initContent()
             'hideLegacyStoreContextSelector' => $this->container->get('prestashop.adapter.multistore_feature')->isUsed(),
         ];
 
-        if ($this->helpLink === false || !empty($this->helpLink)) {
```
Comment (PierreRambaud) [inline controllers/admin/AdminLegacyLayoutController.php]: Why did you remove this.
it can be false or not empty, meaning if you put false you reset the help link and if you put a string it's ok.
I'm afraid of side effect, what was the phpstan error for this one?

File: controllers/admin/AdminQuickAccessesController.php [hunk 5]
```diff
@@ -179,7 +179,7 @@ public function addQuickLink()
             if (method_exists($this->object, 'add') && !$this->object->add()) {
                 $this->errors[] = $this->trans('An error occurred while creating an object.', [], 'Admin.Notifications.Error') .
                     ' <b>' . $this->table . ' (' . Db::getInstance()->getMsgError() . ')</b>';
-            } elseif (($_POST[$this->identifier] = $this->object->id) && $this->postImage($this->object->id) && !count($this->errors) && $this->_redirect) {
+            } elseif (($_POST[$this->identifier] = $this->object->id) && $this->postImage($this->object->id) && !empty($this->errors) && $this->_redirect) {
```
Comment (PierreRambaud) [inline controllers/admin/AdminQuickAccessesController.php]:
```suggestion
            } elseif (($_POST[$this->identifier] = $this->object->id) && $this->postImage($this->object->id) && empty($this->errors) && $this->_redirect) {
```

File: controllers/admin/AdminShopController.php [hunk 6]
```diff
@@ -679,7 +678,7 @@ public function processAdd()
             if (!$object->add()) {
                 $this->errors[] = $this->trans('An error occurred while creating an object.', [], 'Admin.Notifications.Error') .
                     ' <b>' . $this->table . ' (' . Db::getInstance()->getMsgError() . ')</b>';
-            } elseif (($_POST[$this->identifier] = $object->id) && $this->postImage($object->id) && !count($this->errors) && $this->_redirect) {
+            } elseif (($_POST[$this->identifier] = $object->id) && $this->postImage($object->id) && !empty($this->errors) && $this->_redirect) {
```
Comment (PierreRambaud) [inline controllers/admin/AdminShopController.php]:
```suggestion
            } elseif (($_POST[$this->identifier] = $object->id) && $this->postImage($object->id) && empty($this->errors) && $this->_redirect) {
```

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Few comments :wink:

File: classes/Store.php (hunk 1 above)
Comment (Progi1984) [reply classes/Store.php:62]: In `AdminImportController`, the store can be defined as array.

File: controllers/admin/AdminAccessController.php (hunk 2 above)
Comment (Progi1984) [reply controllers/admin/AdminAccessController.php:140]: Yes. But it's a constant... so it should not be a configuration variable.

File: controllers/admin/AdminCustomerThreadsController.php (hunk 3 above)
Comment (Progi1984) [reply controllers/admin/AdminCustomerThreadsController.php:1171]: No because at the line 1169, there is a continue.

File: controllers/admin/AdminLegacyLayoutController.php (hunk 4 above)
Comment (Progi1984) [reply controllers/admin/AdminLegacyLayoutController.php]: Fixed. Sorry.

File: controllers/admin/AdminCustomerThreadsController.php (hunk 3 above)
Comment (PierreRambaud) [reply controllers/admin/AdminCustomerThreadsController.php:1171]: Yes but only if an exception is raised. :thinking:

File: controllers/admin/AdminLegacyLayoutController.php [hunk 7]
```diff
@@ -197,7 +197,7 @@ public function initContent()
...
+        if ($this->helpLink === false ||!empty($this->helpLink)) {
```
Comment (PierreRambaud) [inline controllers/admin/AdminLegacyLayoutController.php]:
```suggestion
        if ($this->helpLink === false || !empty($this->helpLink)) {
```

File: controllers/admin/AdminCustomerThreadsController.php (hunk 3 above)
Comment (Progi1984) [reply controllers/admin/AdminCustomerThreadsController.php:1171]: But at the next step, `fetch_succeed` will be define to `true`.

File: controllers/admin/AdminLegacyLayoutController.php (hunk 7 above)
Comment (Progi1984) [reply controllers/admin/AdminLegacyLayoutController.php]: Fixed

File: controllers/admin/AdminShopController.php (hunk 6 above)
Comment (Progi1984) [reply controllers/admin/AdminShopController.php]: Fixed

File: controllers/admin/AdminQuickAccessesController.php (hunk 5 above)
Comment (Progi1984) [reply controllers/admin/AdminQuickAccessesController.php]: Fixed

File: controllers/admin/AdminCustomerThreadsController.php (hunk 3 above)
Comment (PierreRambaud) [reply controllers/admin/AdminCustomerThreadsController.php:1171]: Oh, I get it!

**Review by PierreRambaud — DISMISSED** [review]

File: classes/controller/AdminController.php [hunk 8]
```diff
@@ -3650,7 +3650,7 @@ public function getFieldsValue($obj)
      * @param string $key Field name
      * @param int|null $id_lang Language id (optional)
      *
-     * @return string|false
+     * @return string|false|mixed
```
Comment (sowbiba) [inline classes/controller/AdminController.php]: return `mixed` isn't enough ?

File: controllers/admin/AdminCartRulesController.php [hunk 9]
```diff
@@ -387,8 +387,8 @@ public function getProductRuleGroupsDisplay($cart_rule)
             $i = 1;
             foreach ($array as $ruleGroupId) {
                 $productRulesArray = [];
-                if (is_array($array = Tools::getValue('product_rule_' . $ruleGroupId)) && count($array)) {
-                    foreach ($array as $ruleId) {
+                if (is_array($arrayPRGroupId = Tools::getValue('product_rule_' . $ruleGroupId)) && count($arrayPRGroupId)) {
```
Comment (sowbiba) [inline controllers/admin/AdminCartRulesController.php]: I would prefer the assignation first before use it in the IF statement. But it's estheticism

**Review by matthieu-rolland — DISMISSED** [review]

File: controllers/admin/AdminCartRulesController.php (hunk 9 above)
Comment (Progi1984) [reply controllers/admin/AdminCartRulesController.php]: Fixed

File: classes/controller/AdminController.php (hunk 8 above)
Comment (Progi1984) [reply classes/controller/AdminController.php]: I add the return value of `Tools::getValue`.

**Comment by Progi1984** [comment]
@PierreRambaud & @matthieu-rolland : Rebased & Changed a line after a comment of @sowbiba

**Review by PierreRambaud — DISMISSED** [review]

**Review by sowbiba — DISMISSED** [review]

**Comment by sowbiba** [comment]
waiting for @boubkerbribri tests ;)

**Review by boubkerbribri — CHANGES_REQUESTED** [review]
Some problems were detected while running UI tests : 
1. BO -> Catalog -> Attribute -> new Value : Empty page displayed.
image

2. BO -> Catalog -> Feature -> new Value : Empty page displayed.
image

**Comment by Progi1984** [comment]
@sowbiba @PierreRambaud I have fixed the bug detected  by @boubkerbribri

**Review by PierreRambaud — APPROVED** [review]

File: classes/controller/AdminController.php [hunk 10]
```diff
@@ -3650,7 +3650,7 @@ public function getFieldsValue($obj)
      * @param string $key Field name
      * @param int|null $id_lang Language id (optional)
      *
-     * @return string|false
+     * @return false|mixed
```
Comment (atomiix) [inline classes/controller/AdminController.php:3653]: I guess `mixed` is enough here
```suggestion
     * @return mixed
```

**Review by atomiix — APPROVED** [review]
Optional comment, still approving ;)

**Review by boubkerbribri — APPROVED** [review]
Tested with UI tests.
It's QA ✔️

File: classes/controller/AdminController.php (hunk 10 above)
Comment (Progi1984) [reply classes/controller/AdminController.php:3653]: Oops.
